src/tools/cephfs/DataScan.cc

```diff
@@ -0,0 +1,1267 @@
+    bool have_backtrace = false;
+    bufferlist parent_bl;
+    int r = data_io.getxattr(oid, "parent", parent_bl);
+    if (r == -ENODATA) {
+      dout(10) << "No backtrace on '" << oid << "'" << dendl;
+    } else if (r < 0) {
+      // TODO: accumulate these errors in a structure we can output
+      dout(4) << "Unexpected error on '" << oid << "': " << cpp_strerror(r) << dendl;
+      continue;
+    }
+
+    // Deserialize backtrace
+    inode_backtrace_t backtrace;
+    if (parent_bl.length()) {
+      try {
+        bufferlist::iterator q = parent_bl.begin();
+        backtrace.decode(q);
+        have_backtrace = true;
+      } catch (buffer::error &e) {
+        dout(4) << "Corrupt backtrace on '" << oid << "': " << e << dendl;
+      }
+    }
+
+    // Santity checking backtrace ino against object name
+    if (have_backtrace && backtrace.ino != obj_name_ino) {
+      dout(4) << "Backtrace ino 0x" << std::hex << backtrace.ino
+        << " doesn't match object name ino 0x" << obj_name_ino
+        << std::dec << dendl;
+      have_backtrace = false;
+    }
+
+    // Read size
+    uint64_t obj_size = 0;
+    time_t obj_mtime = 0;
+    r = data_io.stat(oid, &obj_size, &obj_mtime);
+    if (r != 0) {
+      dout(4) << "Cannot stat '" << oid << "': skipping" << dendl;
+      continue;
+    }
+
+    // Read accumulated size from scan_extents phase
+    uint64_t file_size = 0;
+    uint32_t chunk_size = g_default_file_layout.fl_object_size;
+    bufferlist scan_size_bl;
+    r = data_io.getxattr(oid, "scan_size", scan_size_bl);
+    if (r >= 0) {
+      try {
+        bufferlist::iterator scan_size_bl_iter = scan_size_bl.begin();
+        uint64_t max_obj_id;
+        uint64_t max_obj_size;
+        ::decode(max_obj_id, scan_size_bl_iter);
+        ::decode(max_obj_size, scan_size_bl_iter);
+
+        if (max_obj_id > 0) {
+          // If there are more objects, and this object is a power of two
+          // that is greater than the size of the last object, then make
+          // an educated guess that this object's size is probably the chunk
+          // size.
+        
+          // TODO: provide a means for the user to specify the layout, as
+          // for e.g. striped layouts there is no way we can infer it.
+          if ((obj_size & (obj_size - 1)) == 0 && obj_size >= max_obj_size) {
+            chunk_size = obj_size;
+          } else {
+            chunk_size = g_default_file_layout.fl_object_size;
+          }
+          file_size = chunk_size * max_obj_id + max_obj_size;
+        } else {
+          file_size = obj_size;
+        }
+      } catch (const buffer::error &err) {
+        dout(4) << "Invalid size attr on '" << oid << "'" << dendl;
+        file_size = obj_size;
+      }
+    } else {
+      file_size = obj_size;
+    }
+
+    // Read accumulated mtime from scan_extents phase
+    time_t file_mtime = 0;
+    bufferlist scan_mtime_bl;
+    r = data_io.getxattr(oid, "scan_mtime", scan_mtime_bl);
```

Comment:
As far as I can tell all these recover-time accesses to the zeroth file object should be combined in one op; we're going to be generating enough traffic here without multiplying it.